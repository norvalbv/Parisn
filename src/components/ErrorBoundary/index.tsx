import React, { ReactElement, useMemo } from 'react';
// React error boundaries do not work with functional components out of the box - hence the use of the package below.
import { ErrorBoundary as ErrorBounds } from 'react-error-boundary';

export type ErrorBoundaryProps = {
  children?: JSX.Element;
  errorMessage?: string;
  // Typically used for resetting state or to re-fetch data.
  onClick?: () => void;
  className?: string;
  // Used for automatically specifiying when the error boundary should reload.
  reset?: {
    resetKeys: string[];
    onReset: () => void;
  };
};

const ErrorBoundary = ({
  children,
  errorMessage,
  onClick = (): void => window.location.reload(),
  className = '',
  reset,
}: ErrorBoundaryProps): ReactElement => {
  const memoizedFallback = useMemo(
    () => (
      <div
        role="alert"
        className={`flex items-center gap-4 rounded bg-red-400 px-4 py-2 ${className}`}
      >
        <b>
          {errorMessage ||
            'An error occured! Please try again or contact support if the error persists'}
        </b>
        {onClick && (
          <button
            onClick={(): void => onClick()}
            type="button"
            className="rounded border border-gray-500 px-4 py-2 hover:underline"
          >
            Try again
          </button>
        )}
      </div>
    ),
    [className, errorMessage, onClick]
  );

  return children ? (
    <ErrorBounds
      fallbackRender={(): JSX.Element => memoizedFallback}
      resetKeys={reset?.resetKeys}
      onReset={(): void => reset?.onReset()}
    >
      {children}
    </ErrorBounds>
  ) : (
    memoizedFallback
  );
};

export default ErrorBoundary;
