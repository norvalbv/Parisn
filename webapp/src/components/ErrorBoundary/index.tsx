import React, { ReactElement, useMemo } from 'react';
// React error boundaries do not work with functional components out of the box - hence the use of the package below.
import { ErrorBoundary as ErrorBounds } from 'react-error-boundary';

export interface ErrorBoundaryProps {
  children: JSX.Element;
  errorMessage?: string;
  // Typically used for resetting state or to re-fetch data.
  onClick?: () => void;
  className?: string;
  // Used for automatically specifiying when the error boundary should reload.
  reset?: {
    resetKeys: string[];
    onReset: () => void;
  };
}

const ErrorBoundary = ({
  children,
  errorMessage,
  onClick = () => window.location.reload(),
  className = '',
  reset,
}: ErrorBoundaryProps): ReactElement => {
  const memoizedFallback = useMemo(
    () => (
      <div
        role="alert"
        className={`bg-red-400 rounded py-2 px-4 flex items-center gap-4 ${className}`}
      >
        <b>
          {errorMessage ||
            'An error occured! Please try again or contact support if the error persists'}
        </b>
        {onClick && (
          <button
            onClick={(): void => onClick()}
            type="button"
            className="border border-gray-500 rounded px-4 py-2 hover:underline"
          >
            Try again
          </button>
        )}
      </div>
    ),
    [className, errorMessage, onClick]
  );

  return (
    <ErrorBounds
      fallbackRender={(): JSX.Element => memoizedFallback}
      resetKeys={reset?.resetKeys}
      onReset={(): void => reset?.onReset()}
    >
      {children}
    </ErrorBounds>
  );
};

export default ErrorBoundary;
