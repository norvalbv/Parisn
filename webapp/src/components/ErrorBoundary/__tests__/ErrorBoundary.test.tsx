import React, { FC, ReactElement } from 'react';
import renderer from 'react-test-renderer';

import ErrorBoundary, { ErrorBoundaryProps } from '..';

const ErrorBoundaryWithAllProps: FC<Required<ErrorBoundaryProps>> = ErrorBoundary;

const Child = (): ReactElement => {
  throw new Error();
};

describe('<ErrorBoundary />', () => {
  test('should not render, the div inside should throw an error.', () => {
    const tree = renderer
      .create(
        <ErrorBoundary>
          <div>Child component</div>
        </ErrorBoundary>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render if error is thrown from child', () => {
    const tree = renderer
      .create(
        <ErrorBoundary>
          <Child />
        </ErrorBoundary>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render with all props', () => {
    const tree = renderer
      .create(
        <ErrorBoundaryWithAllProps
          className="m-4 p-4 font-light"
          onClick={jest.fn()}
          errorMessage="Error Message!"
          reset={{ onReset: jest.fn(), resetKeys: ['Reset Key'] }}
        >
          <Child />
        </ErrorBoundaryWithAllProps>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
