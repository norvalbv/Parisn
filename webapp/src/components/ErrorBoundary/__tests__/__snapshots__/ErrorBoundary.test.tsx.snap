// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> should not render, the div inside should throw an error. 1`] = `
<div>
  Child component
</div>
`;

exports[`<ErrorBoundary /> should render if error is thrown from child 1`] = `
<div
  className="bg-red-400 rounded py-2 px-4 flex items-center gap-4 "
  role="alert"
>
  <b>
    An error occured! Please try again or contact support if the error persists
  </b>
  <button
    className="border border-gray-500 rounded px-4 py-2 hover:underline"
    onClick={[Function]}
    type="button"
  >
    Try again
  </button>
</div>
`;

exports[`<ErrorBoundary /> should render with all props 1`] = `
<div
  className="bg-red-400 rounded py-2 px-4 flex items-center gap-4 p-4 m-4 font-light"
  role="alert"
>
  <b>
    Error Message!
  </b>
  <button
    className="border border-gray-500 rounded px-4 py-2 hover:underline"
    onClick={[Function]}
    type="button"
  >
    Try again
  </button>
</div>
`;
